docs/hawck-inputd.1
.\" Automatically generated by Pandoc 2.8
.\"
.TH "HAWCK-INPUTD" "1" "" "Version 0.7" "Keyboard Input Grabber"
.hy
.SH NAME
.PP
\f[B]hawck-inputd\f[R] \[em] Keyboard Input Grabber
.SH SYNOPSIS
.PP
\f[B]hawck-macrod\f[R] [\f[B]OPTIONS\f[R]]\&... [\f[B]--kbd-device\f[R]
\f[I]kbd\f[R]]\&...
.SH DESCRIPTION
.PP
Listen to keyboard devices and pass whitelisted input over to
hawck-macrod.
.PP
Hawck InputD works in two modes:
.IP "1." 3
Listening on all keyboards
.IP \[bu] 2
The default, in this mode hawck-inputd will grab all keyboard devices
when it starts up.
It will also automatically grab the keyboard input of any keyboards that
are plugged in while running.
.IP "2." 3
Listening on only a few specific keyboards
.IP \[bu] 2
Hawck InputD works in this mode when it is given at least one keyboard
with the \f[B]--kbd-device\f[R], as well as \f[B]--no-hotplug\f[R].
.IP \[bu] 2
Keyboards that were initially provided using \f[B]--kbd-device\f[R] are
not affected by \f[B]--no-hotplug\f[R], they can still be removed and
re-attached.
.SS Options
.TP
\f[B]-h\f[R], \f[B]--help\f[R]
Prints brief usage information.
.TP
\f[B]--no-fork\f[R]
Causes hawck-inputd to not fork/daemonize to the background.
.TP
\f[B]-k\f[R], \f[B]--kbd-device\f[R] \f[I]device\f[R]
Select a keyboard device that should be listened for.
.RS
.PP
This keyboard device should be a device file, these device files are
available in the \f[I]/dev/input/by-path\f[R] and
\f[I]/dev/input/by-id\f[R] directories.
.RE
.TP
\f[B]--udev-event-delay\f[R] \f[I]\[mc]s\f[R]
The delay between keyboard events in \[mc]s.
.RS
.PP
Changing this can be useful if you\[cq]re dealing with a program or
desktop environment that is dropping keys.
.RE
.TP
\f[B]--socket-timeout\f[R] \f[I]ms\f[R]
Time until socket timeout in milliseconds.
.RS
.PP
The socket connection in question is the one between
\f[B]hawck-macrod\f[R] and \f[B]hawck-inputd\f[R].
.RE
.TP
\f[B]-v\f[R], \f[B]--version\f[R]
Prints the current version number.
.SH FILES
.TP
\f[I]/var/lib/hawck-input/keys/\f[R]
contains whitelisted keys in csv format.
.TP
\f[I]/var/lib/hawck-input/kbd.sock\f[R]
Is the socket that InputD will connect to and send key events.
.TP
\f[I]/var/lib/hawck-input/pid\f[R]
Contains the pid of the currently running hawck-inputd daemon.
.SH BUGS
.PP
See GitHub Issues: <https://github.com/snyball/Hawck/issues>
.SH COPYRIGHT
.PP
Copyright (C) Jonas M\[/o]ller 2018
.PP
Provided under the BSD 2-clause license.
.SH SEE ALSO
.PP
\f[B]hawck-macrod(1)\f[R], \f[B]hwk2lua(1)\f[R]


src/hawck-inputd.cpp
#include <string>
#include <fstream>
#include <filesystem>

#include "KBDDaemon.hpp"
#include "Daemon.hpp"
#include "utils.hpp"

#if MESON_COMPILE
#include <hawck_config.h>
#else
#define INPUTD_VERSION "unknown"
#endif

extern "C" {
    #include <signal.h>
    #include <getopt.h>
    #include <syslog.h>
}

using namespace std;
namespace fs = std::filesystem;

static void handleSigPipe(int) {
    // cout << "KBDDaemon aborting due to SIGPIPE" << endl;
    // abort();
}

static int no_fork;

auto varToOption(string opt) {
    replace(opt.begin(), opt.end(), '_', '-');
    return opt;
}

#define VAR_TO_OPTION(_var) varToOption(#_var)

auto numOption(const string& name, int *num) {
    return [num, name](const string& opt) {
               try {
                   *num = stoi(opt);
               } catch (const exception &e) {
                   cout << "--" << name << ": Require an integer" << endl;
                   exit(0);
               }
           };
}

auto strOption(const string&, string *str) {
    return [str](const string& opt) {
               *str = opt;
           };
}

#define NUM_OPTION(_name) {VAR_TO_OPTION(_name), numOption(VAR_TO_OPTION(_name), &(_name))},
#define STR_OPTION(_name) {VAR_TO_OPTION(_name), strOption(VAR_TO_OPTION(_name), &(_name))}

int main(int argc, char *argv[]) {
    signal(SIGPIPE, handleSigPipe);

    string HELP =
        "Usage: hawck-inputd [--udev-event-delay <us>] [--no-fork] [--socket-timeout]\n"
        "                    [--kbd-device <device>] [--no-hotplug]\n"
        "\n"
        "Examples:\n"
        "  Listen on a single device:\n"
        "    hawck-inputd --kbd-device /dev/input/event13 --no-hotplug\n\n"
        "  Listen on multiple devices:\n"
        "    hawck-inputd -k{/dev/input/event13,/dev/input/event15}\n\n"
        "  Listen on all keyboard devices automatically:\n"
        "    hawck-inputd\n\n"
        "Options:\n"
        "  --no-fork           Don't daemonize/fork.\n"
        "  -h, --help          Display this help information.\n"
        "  --version           Display version and exit.\n"
        "  -k, --kbd-device    Add a keyboard to listen to.\n"
        "  --udev-event-delay  Delay between events sent on the udevice in us.\n"
        "  --socket-timeout    Time in milliseconds until timeout on sockets.\n"
        "  --no-hotplug        Only listen to devices that were explicitly added with --kbd-device\n"
    ;

    int no_hotplug = false;
    static struct option long_options[] =
        {
            /* These options set a flag. */
            {"no-fork", no_argument,       &no_fork, 1},
            {"no-hotplug", no_argument,       &no_hotplug, 1},
            {"udev-event-delay", required_argument,       0, 0},
            {"socket-timeout", required_argument,       0, 0},
            {"version", no_argument, 0, 0},
            /* These options don’t set a flag.
               We distinguish them by their indices. */
            {"help",         no_argument,       0, 'h'},
            {"kbd-device",   required_argument, 0, 'k'},
            {"kbd-name",     required_argument, 0, 'n'},
            {0, 0, 0, 0}
        };
    /* getopt_long stores the option index here. */
    int option_index = 0;

    int udev_event_delay = 3800;
    int socket_timeout = 1024;
    vector<string> kbd_names;
    vector<string> kbd_devices;
    unordered_map<string, function<void(const string& opt)>> long_handlers = {
        {"version", [&](const string&) {
                        cout << "Hawck InputD v" INPUTD_VERSION << endl;
                        exit(0);
                    }},
        NUM_OPTION(udev_event_delay)
        NUM_OPTION(socket_timeout)
    };

    do {
        int c = getopt_long(argc, argv, "hk:n:",
                            long_options, &option_index);

        /* Detect the end of the options. */
        if (c == -1)
            break;

        switch (c) {
            case 0: {
                /* If this option set a flag, do nothing else now. */
                if (long_options[option_index].flag != 0)
                    break;
                string name(long_options[option_index].name);
                string arg(optarg ? optarg : "");
                if (long_handlers.find(name) != long_handlers.end())
                    long_handlers[name](arg);
                break;
            }

            case 'k':
                kbd_devices.push_back(string(optarg));
                break;

            // TODO: Implement lookup of keyboard names
            case 'n':
                cout << "Option -n/--kbd-name: Not implemented." << endl;
                break;

            case 'h':
                cout << HELP;
                exit(0);

            case '?':
                /* getopt_long already printed an error message. */
                break;

            default:
                cout << "DEFAULT" << endl;
                abort ();
        }
    } while (true);

    // If no devices were specified, we listen to all of them
    if (kbd_devices.size() == 0) {
        // FIXME: All the "is this a keyboard" detection done in Hawck is very
        //        brittle, and I should probably do a deep-dive into whatever
        //        documentation I can find on this to develop a better method.

        for (auto d : fs::directory_iterator("/dev/input/by-id"))
            if (KBDManager::byIDIsKeyboard(d.path()))
                kbd_devices.push_back(realpath_safe(d.path()));

        for (auto d : fs::directory_iterator("/dev/input/by-path")) {
            string rpath = realpath_safe(d.path());
            bool already_added = find(kbd_devices.begin(), kbd_devices.end(), rpath) != kbd_devices.end();
            if (!already_added && stringStartsWith(pathBasename(d.path()), "platform-i8042") && stringEndsWith(d.path(), "-event-kbd"))
                kbd_devices.push_back(rpath);
        }
    }

    cout << "Starting Hawck InputD v" INPUTD_VERSION " on:" << endl;
    for (const auto& dev : kbd_devices)
        cout << "  - <" << dev << ">" << endl;

    if (!no_fork) {
        cout << "forking ..." << endl;
        daemonize("/tmp/hawck-inputd.log");
    }

    const string pid_file = "/var/lib/hawck-input/pid";
    killPretender(pid_file);

    try {
        KBDDaemon daemon;
        daemon.kbman.setHotplug(!no_hotplug);
        for (const auto& dev : kbd_devices)
            daemon.kbman.addDevice(dev);
        daemon.setEventDelay(udev_event_delay);
        daemon.setSocketTimeout(socket_timeout);
        syslog(LOG_INFO, "Running Hawck InputD ...");
        daemon.run();
    } catch (const SystemError &e) {
        syslog(LOG_CRIT, "Abort due to exception: %s", e.what());
        cout << "Error: " << e.what() << endl;
        e.printBacktrace();
        clearPidFile(pid_file);
        throw;
    }
}
